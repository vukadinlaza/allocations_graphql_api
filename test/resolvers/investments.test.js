const { gql } = require("apollo-server");
const { ObjectId } = require("mongodb");
const { getDB } = require("../../src/mongo");
const { SPV_DEAL_ID } = require("../fixtures/deals");
const {
  WIRED_INVESTMENT_ID,
  PENDING_INVESTMENT_ID,
  WIRED_SPV_INVESTMENT_ID,
} = require("../fixtures/investments");
const { INVESTOR_USER_ID } = require("../fixtures/users");
const { describeWithServer, testError } = require("../setup");

const GET_INVESTMENT = gql`
  query GetInvestment($id: String!) {
    investment(_id: $id) {
      _id
      documents {
        path
      }
      value
      deal {
        _id
      }
      investor {
        _id
      }
    }
  }
`;

const GET_INVESTMENTS_LIST = gql`
  query GetInvestmentsList($pagination: PaginationInput!) {
    investmentsList(pagination: $pagination) {
      count
      investments {
        _id
      }
    }
  }
`;

const CREATE_INVESTMENT = gql`
  mutation CreateInvestment($investment: InvestmentInput!) {
    createInvestment(investment: $investment) {
      status
    }
  }
`;

const UPDATE_INVESTMENT = gql`
  mutation UpdateInvestment($investment: InvestmentInput!) {
    updateInvestment(investment: $investment) {
      status
    }
  }
`;

const DELETE_INVESTMENT = gql`
  mutation DeleteInvestment($_id: String!) {
    deleteInvestment(_id: $_id)
  }
`;

describeWithServer(
  "Investments",
  ({
    executeOperationAsAdmin,
    executeOperationAsFundAdmin,
    executeOperationAsInvestor,
    executeOperationAsLoggedIn,
    executeOperation,
  }) => {
    describe("GET_INVESTMENT", () => {
      test("logged in users can get investments", async () => {
        expect(
          await executeOperationAsLoggedIn({
            query: GET_INVESTMENT,
            variables: { id: WIRED_INVESTMENT_ID },
          })
        ).toMatchSnapshot();
      });

      testError("non logged in users cannot get investments", () => {
        expect(
          executeOperation({
            query: GET_INVESTMENT,
            variables: { id: WIRED_INVESTMENT_ID },
          })
        ).rejects.toMatchSnapshot();
      });
    });

    describe("GET_INVESTMENTS_LIST", () => {
      test("admin users can get a paged list of investments", async () => {
        expect(
          await executeOperationAsAdmin({
            query: GET_INVESTMENTS_LIST,
            variables: { pagination: { currentPage: 0, pagination: 1 } },
          })
        ).toMatchSnapshot();

        expect(
          await executeOperationAsAdmin({
            query: GET_INVESTMENTS_LIST,
            variables: { pagination: { currentPage: 1, pagination: 1 } },
          })
        ).toMatchSnapshot();
      });

      testError(
        "non admin users cannot get a paged list of investments",
        async () => {
          expect(
            await executeOperationAsInvestor({
              query: GET_INVESTMENTS_LIST,
              variables: { pagination: { currentPage: 0, pagination: 1 } },
            })
          ).toMatchSnapshot();
        }
      );
    });

    describe("CREATE_INVESTMENT", () => {
      test("logged in users can make investments", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsLoggedIn({
            query: CREATE_INVESTMENT,
            variables: {
              investment: {
                amount: 1000,
                deal_id: SPV_DEAL_ID,
                user_id: INVESTOR_USER_ID,
                status: "onboarded",
              },
            },
          })
        ).toMatchSnapshot();

        expect(await db.investments.findOne({ amount: 1000 })).toMatchSnapshot({
          _id: expect.any(ObjectId),
          created_at: expect.any(Number),
          invited_at: expect.any(Number),
          onboarded_at: expect.any(Number),
        });
      });

      testError("non logged in users cannot make investments", async () => {
        expect(
          executeOperation({
            query: CREATE_INVESTMENT,
            variables: {
              investment: {
                amount: 1000,
                deal_id: SPV_DEAL_ID,
                user_id: INVESTOR_USER_ID,
                status: "onboarded",
              },
            },
          })
        ).rejects.toMatchSnapshot();
      });
    });

    describe("UPDATE_INVESTMENT", () => {
      test("logged in users can update investments", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsLoggedIn({
            query: UPDATE_INVESTMENT,
            variables: {
              investment: {
                _id: PENDING_INVESTMENT_ID,
                amount: 1000,
                status: "wired",
              },
            },
          })
        ).toMatchSnapshot();

        expect(
          await db.investments.findOne({ _id: ObjectId(PENDING_INVESTMENT_ID) })
        ).toMatchSnapshot({
          updated_at: expect.any(Number),
          wired_at: expect.any(Number),
        });
      });

      testError("non logged in users cannot update investments", async () => {
        const db = await getDB();

        expect(
          executeOperation({
            query: UPDATE_INVESTMENT,
            variables: {
              investment: {
                _id: ObjectId(PENDING_INVESTMENT_ID),
                amount: 1000,
                deal_id: SPV_DEAL_ID,
                user_id: INVESTOR_USER_ID,
                status: "onboarded",
              },
            },
          })
        ).rejects.toMatchSnapshot();

        expect(
          await db.investments.findOne({
            _id: ObjectId(PENDING_INVESTMENT_ID),
          })
        ).toMatchSnapshot();
      });
    });

    describe("DELETE_INVESTMENT", () => {
      test("logged in users can delete investments", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsLoggedIn({
            query: DELETE_INVESTMENT,
            variables: { _id: WIRED_INVESTMENT_ID },
          })
        ).toMatchSnapshot();

        expect(
          await db.investments.findOne({ _id: ObjectId(WIRED_INVESTMENT_ID) })
        ).toMatchSnapshot();
      });

      testError("non logged in users cannot delete investments", async () => {
        const db = await getDB();

        expect(
          executeOperation({
            query: DELETE_INVESTMENT,
            variables: { _id: WIRED_INVESTMENT_ID },
          })
        ).rejects.toMatchSnapshot();

        expect(
          await db.investments.findOne({
            _id: ObjectId(WIRED_SPV_INVESTMENT_ID),
          })
        ).toMatchSnapshot();
      });
    });
  }
);
