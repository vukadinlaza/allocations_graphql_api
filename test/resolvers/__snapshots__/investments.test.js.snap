// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investments CREATE_INVESTMENT logged in users can make investments 1`] = `
Object {
  "data": Object {
    "createInvestment": Object {
      "status": "onboarded",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments CREATE_INVESTMENT logged in users can make investments 2`] = `
Object {
  "_id": Any<ObjectId>,
  "amount": 1000,
  "created_at": Any<Number>,
  "deal_id": "61143ac37db3b3619db1df42",
  "invited_at": Any<Number>,
  "onboarded_at": Any<Number>,
  "organization": "61119f11beeab561a48939d3",
  "status": "onboarded",
  "user_id": "611421d64bf8d10b94e61606",
}
`;

exports[`Investments CREATE_INVESTMENT non logged in users cannot make investments 1`] = `[AuthenticationError: authenicate function catch statement]`;

exports[`Investments DELETE_INVESTMENT logged in users can delete investments 1`] = `
Object {
  "data": Object {
    "deleteInvestment": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments DELETE_INVESTMENT logged in users can delete investments 2`] = `null`;

exports[`Investments DELETE_INVESTMENT non logged in users cannot delete investments 1`] = `[AuthenticationError: authenicate function catch statement]`;

exports[`Investments DELETE_INVESTMENT non logged in users cannot delete investments 2`] = `
Object {
  "_id": "61143bc4aa6f42619d924309",
  "amount": 5000,
  "deal_id": "61143ac37db3b3619db1df42",
  "status": "wired",
  "user_id": "611421d64bf8d10b94e61606",
}
`;

exports[`Investments GET_INVESTMENT logged in users can get investments 1`] = `
Object {
  "data": Object {
    "investment": Object {
      "_id": "6111bb42a4991d5a547352b6",
      "deal": Object {
        "_id": "601c54cbe40e5900230f505d",
      },
      "documents": Array [],
      "investor": Object {
        "_id": "611421d64bf8d10b94e61606",
      },
      "value": 25000,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments GET_INVESTMENT non logged in users cannot get investments 1`] = `[AuthenticationError: authenicate function catch statement]`;

exports[`Investments GET_INVESTMENTS_LIST admin users can get a paged list of investments 1`] = `
Object {
  "data": Object {
    "investmentsList": Object {
      "count": 3,
      "investments": Array [
        Object {
          "_id": "6111bb42a4991d5a547352b6",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments GET_INVESTMENTS_LIST admin users can get a paged list of investments 2`] = `
Object {
  "data": Object {
    "investmentsList": Object {
      "count": 3,
      "investments": Array [
        Object {
          "_id": "6111bbe6ad565e63b254d0df",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments GET_INVESTMENTS_LIST non admin users cannot get a paged list of investments 1`] = `
Object {
  "data": Object {
    "investmentsList": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments UPDATE_INVESTMENT logged in users can update investments 1`] = `
Object {
  "data": Object {
    "updateInvestment": Object {
      "status": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Investments UPDATE_INVESTMENT logged in users can update investments 2`] = `
Object {
  "_id": "6111bbe6ad565e63b254d0df",
  "amount": 1000,
  "deal_id": "601c54cbe40e5900230f505d",
  "status": "wired",
  "updated_at": Any<Number>,
  "user_id": "611421d64bf8d10b94e61606",
  "wired_at": Any<Number>,
}
`;

exports[`Investments UPDATE_INVESTMENT non logged in users cannot update investments 1`] = `[AuthenticationError: authenicate function catch statement]`;

exports[`Investments UPDATE_INVESTMENT non logged in users cannot update investments 2`] = `
Object {
  "_id": "6111bbe6ad565e63b254d0df",
  "amount": 10000,
  "deal_id": "601c54cbe40e5900230f505d",
  "status": "invited",
  "user_id": "611421d64bf8d10b94e61606",
}
`;
